src/g2.rs

```diff
@@ -1551,6 +1654,174 @@ fn test_is_torsion_free() {
+    assert_eq!(point.psi() + generator.psi(), (point + generator).psi());
+    // psi(P) behaves in the same way on the same projective point
+    let mut normalized_point = [G2Affine::identity()];
+    G2Projective::batch_normalize(&[point], &mut normalized_point);
+    let normalized_point = G2Projective::from(normalized_point[0]);
+    assert_eq!(point.psi(), normalized_point.psi());
+    assert_eq!(point.psi2(), normalized_point.psi2());
+}
+
+#[test]
+fn test_clear_cofactor() {
+    // the generator (and the identity) are always on the curve,
+    // even after clearing the cofactor
+    let a = G2Projective::generator();
+    assert!(bool::from(a.clear_cofactor().is_on_curve()));
+    let a = G2Projective::identity();
+    assert!(bool::from(a.clear_cofactor().is_on_curve()));
+
+    // `point` is a random point in the curve
+    let point = G2Projective {
+        x: Fp2 {
+            c0: Fp::from_raw_unchecked([
+                0xee4c8cb7c047eaf2,
+                0x44ca22eee036b604,
+                0x33b3affb2aefe101,
+                0x15d3e45bbafaeb02,
+                0x7bfc2154cd7419a4,
+                0x0a2d0c2b756e5edc,
+            ]),
+            c1: Fp::from_raw_unchecked([
+                0xfc224361029a8777,
+                0x4cbf2baab8740924,
+                0xc5008c6ec6592c89,
+                0xecc2c57b472a9c2d,
+                0x8613eafd9d81ffb1,
+                0x10fe54daa2d3d495,
+            ]),
+        },
+        y: Fp2 {
+            c0: Fp::from_raw_unchecked([
+                0x7de7edc43953b75c,
+                0x58be1d2de35e87dc,
+                0x5731d30b0e337b40,
+                0xbe93b60cfeaae4c9,
+                0x8b22c203764bedca,
+                0x01616c8d1033b771,
+            ]),
+            c1: Fp::from_raw_unchecked([
+                0xea126fe476b5733b,
+                0x85cee68b5dae1652,
+                0x98247779f7272b04,
+                0xa649c8b468c6e808,
+                0xb5b9a62dff0c4e45,
+                0x1555b67fc7bbe73d,
+            ]),
+        },
+        z: Fp2 {
+            c0: Fp::from_raw_unchecked([
+                0x0ef2ddffab187c0a,
+                0x2424522b7d5ecbfc,
+                0xc6f341a3398054f4,
+                0x5523ddf409502df0,
+                0xd55c0b5a88e0dd97,
+                0x066428d704923e52,
+            ]),
+            c1: Fp::from_raw_unchecked([
+                0x538bbe0c95b4878d,
+                0xad04a50379522881,
+                0x6d5c05bf5c12fb64,
+                0x4ce4a069a2d34787,
+                0x59ea6c8d0dffaeaf,
+                0x0d42a083a75bd6f3,
+            ]),
+        },
+    };
+
+    assert!(bool::from(point.is_on_curve()));
+    assert!(!bool::from(G2Affine::from(point).is_torsion_free()));
+    let cleared_point = point.clear_cofactor();
+
+    assert!(bool::from(cleared_point.is_on_curve()));
+    assert!(bool::from(G2Affine::from(cleared_point).is_torsion_free()));
```

Comment:
I'd have loved to do that, but the current API does not allow for scalar mult by a huge number; even more, multiply(&[u8;32]) seems to be "optimized" for Fp elements. What would you like me to do, another double-and-add implementation inside the function, or have multiply just take a slice?

Reply:
We're only concerned about the effect on the q-torsion (since the h-torsion is killed) so you can just use a `Scalar`. If indeed the paper is correct and the effect can be modeled as a multiplication by `3 * (x^2 - 1) * cofactor` then

```
sage: h_eff = 3 * (param^2 - 1) * g2_h(param)
sage: h_eff
209869847837335686905080341498658477663839067235703451875306851526599783796572738804459333109033834234622528588876978987822447936461846631641690358257586228683615991308971558879306463436166481
sage: h_eff % q
915955240610598312289679060287459819519
```

is the scalar that it should be tested against.

Reply:
This test is a bit weird to me, because I would have said instead we're mostly interested in the effect on points outside the q-torsion, but I guess it's better to have it then not!

Reply:
Callers of `clear_cofactor` may include implementations of "hashing to the group" which depend on the specific behavior of this method and how it might affect the q-torsion. You could replace this method with one that only multiplies by the cofactor, and it would break the hash functions that use it and depend on it. In any case, we can use `clear_cofactor` on points without h-torsion (for some reason) so the actual impact on the point is part of the public API and so must have regression tests.

Reply:
This test becomes useful when feature guarding the endomorphism optimizations for the purpose of avoiding patents.